torchdynamo/_eval_frame.c

```diff
@@ -92,20 +92,65 @@ inline static void eval_frame_callback_set(PyObject *obj) {
 #endif
 
 static void ignored(void *obj) {}
-static PyObject *custom_eval_frame(PyFrameObject *frame, int throw_flag);
-static PyObject *custom_eval_frame_run_only(PyFrameObject *frame,
+static PyObject *_custom_eval_frame(PyThreadState *tstate,
+                                    PyFrameObject *frame,
+                                    int throw_flag);
+static PyObject *_custom_eval_frame_run_only(PyFrameObject *frame,
                                             int throw_flag);
+#if PY_VERSION_HEX >= 0x03090000
+static PyObject *custom_eval_frame(PyThreadState *tstate,
+                                   PyFrameObject *frame,
+                                   int throw_flag) {
+  return _custom_eval_frame(tstate, frame, throw_flag);
+}
+static PyObject *custom_eval_frame_run_only(PyThreadState *tstate,
+                                            PyFrameObject *frame,
+                                            int throw_flag) {
+  return _custom_eval_frame_run_only(frame, throw_flag);
+}
+#else
+static PyObject *custom_eval_frame(PyFrameObject *frame, int throw_flag) {
+  PyThreadState *tstate = PyThreadState_GET();
+  return _custom_eval_frame(tstate, frame, throw_flag);
+}
+static PyObject *custom_eval_frame_run_only(PyFrameObject *frame,
+                                            int throw_flag) {
+  return _custom_eval_frame_run_only(frame, throw_flag);
+}
+#endif
+
+inline static PyObject *eval_frame_default(PyFrameObject *frame,
+                                           int throw_flag) {
+  #if PY_VERSION_HEX >= 0x03090000
+  PyThreadState *tstate = PyThreadState_GET();
```

Comment:
`PyThreadState_GET()` is expensive, which is why it is passed in in 3.9+.

We should avoid adding extra calls to it.  I'd suggest refactoring this to take an optional `PyThreadState *tstate`.  Then only call PyThreadState_GET() if `tstate==NULL`.

Most of the callers will already have a tstate pointer.